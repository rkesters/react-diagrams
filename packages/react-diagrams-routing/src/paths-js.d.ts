declare module 'paths-js/path' {
	export interface Instruction {
		params: number[];
		command: 'default' | 'V' | 'H' | 'A' | 'C' | 'Z';
	}

	export interface Point {
		x: number;
		y: number;
	}

	export interface CurveTo {
		x1: number;
		y1: number;
		x2: number;
		y2: number;
		x: number;
		y: number;
	}

	export interface SmoothCurveTo {
		x2: number;
		y2: number;
		x: number;
		y: number;
	}

	export interface QCurveTo {
		x1: number;
		y1: number;
		x: number;
		y: number;
	}

	export interface Arc {
		rx: number;
		ry: number;
		xrot: number;
		largeArcFlag: number;
		sweepFlag: number;
		x: number;
		y: number;
	}

	export interface Translate {
		dx: number;
		dy: number;
	}

	export interface Rotate {
		angle: number;
		rx: number;
		ry: number;
	}

	export interface Scale {
		sx: number;
		sy: number;
	}

	export interface Shear {

	}

	export interface IPath {
		moveto(point: Point): IPath;
		moveto(x: number, y: number): IPath;
		moveto(x: number | Point, y?: number): IPath;

		lineto(point: Point): IPath;
		lineto(x: number, y: number): IPath;
		lineto(x: number | Point, y?: number): IPath;

		hlineto(point: Point): IPath;
		hlineto(x: number, y: number): IPath;
		hlineto(x: number | Point, y?: number): IPath;

		vlineto(point: Point): IPath;
		vlineto(x: number, y: number): IPath;
		vlineto(x: number | Point, y?: number): IPath;

		closepath(): IPath;

		curveto(curve: CurveTo): IPath;
		curveto(x1: number, y1: number, x2: number, y2: number, x: number): IPath;
		curveto(x1: number | CurveTo, y1?: number, x2?: number, y2?: number, x?: number): IPath;

		smoothcurveto(point: SmoothCurveTo): IPath;
		smoothcurveto(x2: number, y2: number, x: number): IPath;
		smoothcurveto(x2: number | SmoothCurveTo, y2?: number, x?: number): IPath;

		qcurveto(point: QCurveTo): IPath;
		qcurveto(x1: number, y1: number, x: number): IPath;
		qcurveto(x1: number | QCurveTo, y1?: number, x?: number): IPath;

		smoothqcurveto(point: Point): IPath;
		smoothqcurveto(x: number, y: number): IPath;
		smoothqcurveto(x: number | Point, y?: number): IPath;

		arc(point: Arc): IPath;
		arc(rx: number, ry: number, xrot: number, largeArcFlag: number, sweepFlag: number, x: number, y: number): IPath;
		arc(
			rx: number | Arc,
			ry?: number,
			xrot?: number,
			largeArcFlag?: number,
			sweepFlag?: number,
			x?: number,
			y?: number
		): IPath;

		translate(point: Translate): IPath;
		translate(dx: number, dy: number): IPath;
		translate(dx: number | Translate, dy?: number): IPath;

		rotate(point: Rotate): IPath;
		rotate(angle: number, rx: number, ry: number): IPath;
		rotate(angle: number | Rotate, rx: number, ry: number): IPath;

		scale(scaleBy: Scale): IPath;
		scale(sx: number, sy: number): IPath;
		scale(sx: number | Scale, sy?: number): IPath;

		shearX(angle: Scale): IPath;
		shearX(angle: number): IPath;
		shearX(angle: Scale | number): IPath;

		shearY(angle: Scale): IPath;
		shearY(angle: number): IPath;
		shearY(angle: Scale | number): IPath;

		print(): string;
		toString(): string;
		points() : [number, number][];
		instructions() : Instruction[];

		connect(path: IPath): IPath;
	}

	export default function Path(init?: Instruction[]): IPath;
}
