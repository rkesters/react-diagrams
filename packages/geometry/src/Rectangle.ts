import { isNumber } from 'lodash';
import { Point } from './Point';
import { Polygon } from './Polygon';

export class Rectangle extends Polygon {
	constructor(tl: Point, tr: Point, br: Point, bl: Point);
	constructor(position: Point, width: number, height: number);
	constructor(x?: number, y?: number, width?: number, height?: number);

	constructor(a: Point | number = 0, b: Point | number  = 0, c: Point | number  = 0, d: Point | number  = 0) {
		if (Point.isa(a) && Point.isa(b) && Point.isa(c) && Point.isa(d)) {
			super([a, b, c, d]);
			return;
		} else if (Point.isa(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super([a, new Point(a.x + b, a.y), new Point(a.x + b, a.y + c), new Point(a.x, a.y + c)]);
			return;
		} else if (isNumber(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super(Rectangle.pointsFromBounds(a, b, c, d));
			return;
		}

		throw new Error(`Rectangle call with invaild arguements: ${arguments.length}`);
	}

	static pointsFromBounds(x: number, y: number, width: number, height: number): Point[] {
		return [new Point(x, y), new Point(x + width, y), new Point(x + width, y + height), new Point(x, y + height)];
	}

	updateDimensions(x: number, y: number, width: number, height: number) {
		this.points = Rectangle.pointsFromBounds(x, y, width, height);
	}

	setPoints(points: Point[]) {
		if (points.length !== 4) {
			throw 'Rectangles must always have 4 points';
		}
		super.setPoints(points);
	}

	containsPoint(point: Point) {
		const tl = this.getTopLeft();
		const br = this.getBottomRight();

		return point.x >= tl.x && point.x <= br.x && point.y >= tl.y && point.y <= br.y;
	}

	getWidth(): number {
		return Math.sqrt(
			Math.pow(this.getTopLeft().x - this.getTopRight().x, 2) + Math.pow(this.getTopLeft().y - this.getTopRight().y, 2)
		);
	}

	getHeight(): number {
		return Math.sqrt(
			Math.pow(this.getBottomLeft().x - this.getTopLeft().x, 2) +
				Math.pow(this.getBottomLeft().y - this.getTopLeft().y, 2)
		);
	}

	getTopMiddle(): Point {
		return Point.middlePoint(this.getTopLeft(), this.getTopRight());
	}

	getBottomMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getBottomRight());
	}

	getLeftMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getTopLeft());
	}

	getRightMiddle(): Point {
		return Point.middlePoint(this.getBottomRight(), this.getTopRight());
	}

	getTopLeft(): Point {
		return this.points[0];
	}

	getTopRight(): Point {
		return this.points[1];
	}

	getBottomRight(): Point {
		return this.points[2];
	}

	getBottomLeft(): Point {
		return this.points[3];
	}
}
