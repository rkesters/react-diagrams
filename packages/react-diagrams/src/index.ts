import {
	DefaultDiagramState,
	DiagramEngine,
	DiagramModel,
	LinkLayerFactory,
	NodeLayerFactory
} from '@projectstorm/react-diagrams-core';
import {
	DefaultLabelFactory,
	DefaultLinkFactory,
	DefaultNodeFactory,
	DefaultPortFactory
} from '@projectstorm/react-diagrams-defaults';
import { PathFindingLinkFactory } from '@projectstorm/react-diagrams-routing';
import { SelectionBoxLayerFactory, CanvasEngineOptions } from '@projectstorm/react-canvas-core';

export * from '@projectstorm/react-diagrams-core';
export * from '@projectstorm/react-diagrams-defaults';
export * from '@projectstorm/react-diagrams-routing';

/**
 * Construct an engine with the defaults installed
 */
export default (model: DiagramModel, options: CanvasEngineOptions = {}): DiagramEngine => {
	const engine = new DiagramEngine(model,options);

	// register model factories
	engine.getLayerFactories().registerFactory(new NodeLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new LinkLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new SelectionBoxLayerFactory(engine));

	engine.getLabelFactories().registerFactory(new DefaultLabelFactory(engine));
	engine.getNodeFactories().registerFactory(new DefaultNodeFactory(engine)); // i cant figure out why
	engine.getLinkFactories().registerFactory(new DefaultLinkFactory(engine));
	engine.getLinkFactories().registerFactory(new PathFindingLinkFactory(engine));
	engine.getPortFactories().registerFactory(new DefaultPortFactory(engine));

	// register the default interaction behaviours
	engine.getStateMachine().pushState(new DefaultDiagramState(engine));
	return engine;
};
