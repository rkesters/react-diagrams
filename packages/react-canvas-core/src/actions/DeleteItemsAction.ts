import * as _ from 'lodash';
import { KeyboardEvent } from 'react';
import { Action, ActionEvent, InputType } from '../core-actions/Action';

export interface KeyModifiers {
		ctrlKey?: boolean;
		shiftKey?: boolean;
		altKey?: boolean;
		metaKey?: boolean;
}
export interface DeleteItemsActionOptions {
	 /** @deprecated */
	keyCodes?: number[];
	keys?: string[]
	modifiers?: KeyModifiers
}

function keyTest  (keys: string[] , keyCodes:number[], modifiers: KeyModifiers, key: string, keyCode: number, ctrlKey: boolean, shiftKey: boolean, altKey: boolean, metaKey: boolean) {
	return (keys.includes(key) && _.isEqual({ ctrlKey, shiftKey, altKey, metaKey }, modifiers))
}

function keyCodeTest  (keys: string[] , keyCodes:number[], modifiers: KeyModifiers,key: string, keyCode: number,  ctrlKey: boolean, shiftKey: boolean, altKey: boolean, metaKey: boolean)  {
return (keyCodes.includes(keyCode) && _.isEqual({ ctrlKey, shiftKey, altKey, metaKey }, modifiers));
}
/**
 * Deletes all selected items
 */
export class DeleteItemsAction extends Action {

	#deleteOptions: DeleteItemsActionOptions;

	static #defaultOptions = { keys: ['Backspace', 'Delete']  }

	public get deleteOptions(): Readonly< DeleteItemsActionOptions> {
		 return this.#deleteOptions;
	}

	constructor(options: DeleteItemsActionOptions = {}) {

		super({
			type: InputType.KEY_DOWN,
			fire: (event: ActionEvent<KeyboardEvent>) => {
				const keyCodes = this.#deleteOptions.keyCodes || [46, 8];
				const keys = this.#deleteOptions.keys ?? ['Backspace', 'Delete'];
				const test = _.isEmpty(this.#deleteOptions.keyCodes) ? keyTest : keyCodeTest;
				const modifiers = {
					ctrlKey: false,
					shiftKey: false,
					altKey: false,
					metaKey: false,
					...(this.#deleteOptions.modifiers ?? {})
				};
				const { key, keyCode, ctrlKey, shiftKey, altKey, metaKey } = event.event;
				if (test(keys,keyCodes, modifiers, key, keyCode, ctrlKey, shiftKey, altKey, metaKey)) {
					_.forEach(this.engine.getModel().getSelectedEntities(), (model) => {
						// only delete items which are not locked
						if (!model.isLocked()) {
							model.remove();
						}
					});
					this.engine.repaintCanvas();
				}
			}
		});
		this.#deleteOptions = _.isEmpty( options) ? _.defaults (options,DeleteItemsAction.#defaultOptions) : {...options};

	}
}
