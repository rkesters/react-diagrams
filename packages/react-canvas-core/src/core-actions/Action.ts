import { MouseEvent, KeyboardEvent, WheelEvent, TouchEvent, SyntheticEvent } from 'react';
import { Toolkit } from '../Toolkit';
import { CanvasEngine } from '../CanvasEngine';
import { BaseModel } from '../core-models/BaseModel';

export enum InputType {
	MOUSE_DOWN = 'mouse-down',
	MOUSE_UP = 'mouse-up',
	MOUSE_MOVE = 'mouse-move',
	MOUSE_WHEEL = 'mouse-wheel',
	KEY_DOWN = 'key-down',
	KEY_UP = 'key-up',
	TOUCH_START = 'touch-start',
	TOUCH_END = 'touch-end',
	TOUCH_MOVE = 'touch-move'
}

export interface Mapping {
	[InputType.MOUSE_DOWN]: MouseEvent;
	[InputType.MOUSE_UP]: MouseEvent;
	[InputType.MOUSE_MOVE]: MouseEvent;
	[InputType.MOUSE_WHEEL]: WheelEvent;
	[InputType.KEY_DOWN]: KeyboardEvent;
	[InputType.KEY_UP]: KeyboardEvent;
	[InputType.TOUCH_START]: TouchEvent;
	[InputType.TOUCH_END]: TouchEvent;
	[InputType.TOUCH_MOVE]: TouchEvent;
}

export interface ActionEvent<Event extends SyntheticEvent = SyntheticEvent, Model extends BaseModel = BaseModel> {
	event: Event;
	model?: Model;
}

export interface ActionOptionsMouse{
	type: InputType.MOUSE_DOWN | InputType.MOUSE_MOVE | InputType.MOUSE_UP;
	fire: (event: ActionEvent<MouseEvent>) => void;
}
export interface ActionOptionsWheel{
type: InputType.MOUSE_WHEEL;
fire: (event: ActionEvent<WheelEvent>) => void;
}

export interface ActionOptionsKeyboard{
type: InputType.KEY_DOWN | InputType.KEY_UP;
fire: (event: ActionEvent<KeyboardEvent>) => void;
}
export interface ActionOptionsTouch{
type: InputType.TOUCH_END | InputType.TOUCH_MOVE | InputType.TOUCH_START;
fire: (event: ActionEvent<TouchEvent>) => void;
}

type ActionOptions = ActionOptionsMouse | ActionOptionsWheel | ActionOptionsKeyboard  | ActionOptionsTouch

export class Action<T extends CanvasEngine = CanvasEngine> {
	options: ActionOptions;
	id: string;
	engine: T | undefined;

	constructor(options: ActionOptions) {
		this.options = options;
		this.id = Toolkit.UID();
	}

	setEngine(engine: T) {
		this.engine = engine;
	}
}
