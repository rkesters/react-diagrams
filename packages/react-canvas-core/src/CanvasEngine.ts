import { debounce } from 'lodash';
import { CanvasModel } from './entities/canvas/CanvasModel';
import { FactoryBank } from './core/FactoryBank';
import { AbstractReactFactory } from './core/AbstractReactFactory';
import { LayerModel } from './entities/layer/LayerModel';
import { BaseListener, BaseObserver } from './core/BaseObserver';
import { MouseEvent } from 'react';
import { BaseModel } from './core-models/BaseModel';
import { Point } from '@projectstorm/geometry';
import { ActionEventBus } from './core-actions/ActionEventBus';
import { ZoomCanvasAction } from './actions/ZoomCanvasAction';
import { DeleteItemsAction } from './actions/DeleteItemsAction';
import { StateMachine } from './core-state/StateMachine';
export interface CanvasEngineListener extends BaseListener {
	canvasReady(): void;

	repaintCanvas(): void;

	rendered(): void;
}

/**
 * Defines the CanvasEngine options
 */
export interface CanvasEngineOptions {
	registerDefaultDeleteItemsAction?: boolean;
	registerDefaultZoomCanvasAction?: boolean;
	/**
	 * Defines the debounce wait time in milliseconds if > 0
	 */
	repaintDebounceMs?: number;
}

export class CanvasEngine<
	L extends CanvasEngineListener = CanvasEngineListener,
	M extends CanvasModel = CanvasModel
> extends BaseObserver<L> {
	protected model: M ;
	protected layerFactories: FactoryBank<AbstractReactFactory<LayerModel>>;
	protected canvas: HTMLDivElement | undefined;
	protected eventBus: ActionEventBus;
	protected stateMachine: StateMachine;
	protected options: CanvasEngineOptions;

	constructor(model: M, options: CanvasEngineOptions = {}) {
		super();
		this.model = model;
		this.eventBus = new ActionEventBus(this);
		this.stateMachine = new StateMachine(this);
		this.layerFactories = new FactoryBank();
		this.registerFactoryBank(this.layerFactories);

		/**
		 * Overrides the standard options with the possible given options
		 */
		this.options = {
			registerDefaultDeleteItemsAction: true,
			registerDefaultZoomCanvasAction: true,
			repaintDebounceMs: 0,
			...options
		};
		if (this.options.registerDefaultZoomCanvasAction === true) {
			this.eventBus.registerAction(new ZoomCanvasAction({}, this));
		}
		if (this.options.registerDefaultDeleteItemsAction === true) {
			this.eventBus.registerAction(new DeleteItemsAction({}, this));
		}
	}

	getStateMachine() {
		return this.stateMachine;
	}

	getRelativeMousePoint(event: { clientX: number; clientY: number }): Point {
		const point = this.getRelativePoint(event.clientX, event.clientY);
		return new Point(
			(point.x - (this.model?.getOffsetX() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0),
			(point.y - (this.model?.getOffsetY() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0)
		);
	}

	getRelativePoint(x:number, y: number): Point {
		const canvasRect = this.canvas?.getBoundingClientRect();
		if (!canvasRect) {
			throw new Error(`Can't get relative point because canvas has not been defined.`)
		}
		return new Point(x - canvasRect.left, y - canvasRect.top);
	}

	registerFactoryBank(factory: FactoryBank<any>) {
		factory.registerListener({
			factoryAdded: (event) => {
				event.factory.setDiagramEngine(this);
			},
			factoryRemoved: (event) => {
				event.factory.setDiagramEngine(null);
			}, eventWillFire: ():void => {} , eventDidFire: ():void => {}
		});
	}

	getActionEventBus() {
		return this.eventBus;
	}

	getLayerFactories() {
		return this.layerFactories;
	}

	getFactoryForLayer<F extends AbstractReactFactory<LayerModel>>(layer: LayerModel | string | undefined) {
		if (!layer) { return;}
		if (typeof layer === 'string') {
			return this.layerFactories.getFactory(layer);
		}
		return this.layerFactories.getFactory(layer.getType());
	}

	setModel(model: M) {
		this.model = model;
		if (this.canvas) {
			requestAnimationFrame(() => {
				this.repaintCanvas();
			});
		}
	}

	getModel(): M {
		return this.model;
	}

	repaintCanvas(promise: true): Promise<any>;
	repaintCanvas(): void;
	repaintCanvas(promise?: true): Promise<any> | void {
		const repaintDebounceMs  = this.options.repaintDebounceMs ?? 0;

		/**
		 * The actual repaint function
		 */
		const repaint = () => {
			this.iterateListeners((listener) => {
				if (listener.repaintCanvas) {
					listener.repaintCanvas();
				}
			});
		};

		// if the `repaintDebounceMs` option is > 0, then apply the debounce
		let repaintFn = repaint;

		if (repaintDebounceMs > 0) {
			repaintFn = debounce(repaint, repaintDebounceMs);
		}

		if (promise) {
			return new Promise<void>((resolve) => {
				const l = this.registerListener({
					rendered: () => {
						resolve();
						l.deregister();
					}
				} as L);
				repaintFn();
			});
		}

		repaintFn();
	}

	setCanvas(canvas?: HTMLDivElement | null) {
		if (this.canvas !== canvas) {
			this.canvas = canvas ?? undefined;
			if (canvas) {
				this.fireEvent(undefined, 'canvasReady');
			}
		}
	}

	getCanvas() {
		return this.canvas;
	}

	getMouseElement(event: MouseEvent): BaseModel<any> | null | undefined  {
		return null;
	}

	zoomToFit() {
		if (!this.canvas) {
			throw new Error(`Can not zoom to fit because canvas is not defined`)
		}
		const xFactor = this.canvas.clientWidth / this.canvas.scrollWidth;
		const yFactor = this.canvas.clientHeight / this.canvas.scrollHeight;
		const zoomFactor = xFactor < yFactor ? xFactor : yFactor;

		this.model?.setZoomLevel((this.model?.getZoomLevel() ?? 1) * zoomFactor);
		this.model?.setOffset(0, 0);
		this.repaintCanvas();
	}
}
