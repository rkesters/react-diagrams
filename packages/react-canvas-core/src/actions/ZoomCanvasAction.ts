import { WheelEvent } from 'react';
import { CanvasEngine } from '../CanvasEngine';
import { Action, ActionEvent, InputType } from '../core-actions/Action';

export interface ZoomCanvasActionOptions {
	inverseZoom?: boolean;
}

export class ZoomCanvasAction extends Action {
	constructor(options: ZoomCanvasActionOptions = {}, engine: CanvasEngine ) {
		super({
			type: InputType.MOUSE_WHEEL,
			fire: (actionEvent: ActionEvent<WheelEvent>) => {
				const { event } = actionEvent;
				// we can block layer rendering because we are only targeting the transforms
				for (let layer of (this.engine.getModel()?.getLayers() ?? [])) {
					layer.allowRepaint(false);
				}

				const model = this.engine.getModel();
				event.stopPropagation();
				const oldZoomFactor = (this.engine.getModel()?.getZoomLevel() ?? 1) / 100;
				let scrollDelta = options.inverseZoom ? -event.deltaY : event.deltaY;
				//check if it is pinch gesture
				if (event.ctrlKey && scrollDelta % 1 !== 0) {
					/*
						Chrome and Firefox sends wheel event with deltaY that
						have fractional part, also `ctrlKey` prop of the event is true
						though ctrl isn't pressed
					*/
					scrollDelta /= 3;
				} else {
					scrollDelta /= 60;
				}
				if ((model?.getZoomLevel() ?? 1) + scrollDelta > 10) {
					model?.setZoomLevel(model.getZoomLevel() + scrollDelta);
				}

				const zoomFactor = (model?.getZoomLevel() ?? 1) / 100;

				const boundingRect = event.currentTarget.getBoundingClientRect();
				const clientWidth = boundingRect.width;
				const clientHeight = boundingRect.height;
				// compute difference between rect before and after scroll
				const widthDiff = clientWidth * zoomFactor - clientWidth * oldZoomFactor;
				const heightDiff = clientHeight * zoomFactor - clientHeight * oldZoomFactor;
				// compute mouse coords relative to canvas
				const clientX = event.clientX - boundingRect.left;
				const clientY = event.clientY - boundingRect.top;

				// compute width and height increment factor
				const xFactor = (clientX - (model?.getOffsetX() ?? 0)) / oldZoomFactor / clientWidth;
				const yFactor = (clientY - (model?.getOffsetY() ?? 0)) / oldZoomFactor / clientHeight;

				model?.setOffset((model?.getOffsetX() ?? 0) - widthDiff * xFactor, (model?.getOffsetY() ?? 0) - heightDiff * yFactor);
				this.engine.repaintCanvas();

				// re-enable rendering
				for (let layer of (this.engine.getModel()?.getLayers() ?? [])) {
					layer.allowRepaint(true);
				}
			}
		}, engine);
	}
}
