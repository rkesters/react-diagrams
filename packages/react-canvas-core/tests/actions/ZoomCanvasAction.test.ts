import { defaults } from 'lodash';
import { WheelEvent } from 'react';
import { mocked } from 'ts-jest/utils';
import { ZoomCanvasAction, ZoomCanvasActionOptions } from '../../src/actions/ZoomCanvasAction';
import { CanvasEngine } from '../../src/CanvasEngine';
import { ActionEvent } from '../../src/core-actions/Action';
import { CanvasModel } from '../../src/entities/canvas/CanvasModel';

interface zoomEventOptions {
	deltaY: number;
	clientX: number;
	clientY: number;
	ctrlKey?: boolean;
	boundingBox: { width: number; height: number; left: number; top: number };
}

const generateZoomEvent = (opts: zoomEventOptions): ActionEvent<WheelEvent> => {
	const event: WheelEvent = {
		deltaY: opts.deltaY,
		clientX: opts.clientX,
		clientY: opts.clientY,

		currentTarget: {
			getBoundingClientRect: jest.fn().mockImplementation((): DOMRect => {
				return { ...opts.boundingBox } as DOMRect;
			})
		},
		stopPropagation: jest.fn()
	} as any;
	return { event } as ActionEvent<WheelEvent>;
};
interface TestCases {
	options?: ZoomCanvasActionOptions;
	events: zoomEventOptions[];
	name: string;
}

const testCases: TestCases[] = [
	{
		name: 'inverseZoom True',
		options: { inverseZoom: true },
		events: [{ deltaY: 20, clientX: 10, clientY: 10, boundingBox: { width: 500, height: 500, left: 0, top: 0 } }]
	},
	{
		name: 'inverseZoom false',
		options: { inverseZoom: false },
		events: [{ deltaY: 20, clientX: 10, clientY: 10, boundingBox: { width: 500, height: 500, left: 0, top: 0 } }]
	},
	{
		name: 'inverseZoom undefined',
		options: {},
		events: [{ deltaY: 20, clientX: 10, clientY: 10, boundingBox: { width: 500, height: 500, left: 0, top: 0 } }]
	},
	{
		name: 'Options undefined',
		events: [{ deltaY: 20, clientX: 10, clientY: 10, boundingBox: { width: 500, height: 500, left: 0, top: 0 } }]
	},
	{
		name: 'inverseZoom True, ctrlKey true',
		options: { inverseZoom: true },
		events: [
			{
				deltaY: 20.123,
				ctrlKey: true,
				clientX: 10,
				clientY: 10,
				boundingBox: { width: 500, height: 500, left: 0, top: 0 }
			}
		]
	},
	{
		name: 'inverseZoom false',
		options: { inverseZoom: false },
		events: [
			{
				deltaY: 20.123,
				ctrlKey: true,
				clientX: 10,
				clientY: 10,
				boundingBox: { width: 500, height: 500, left: 0, top: 0 }
			}
		]
	},
	{
		name: 'inverseZoom undefined',
		options: {},
		events: [
			{
				deltaY: 20.123,
				ctrlKey: true,
				clientX: 10,
				clientY: 10,
				boundingBox: { width: 500, height: 500, left: 0, top: 0 }
			}
		]
	},
	{
		name: 'Options undefined',
		events: [
			{
				deltaY: 20.123,
				ctrlKey: true,
				clientX: 10,
				clientY: 10,
				boundingBox: { width: 500, height: 500, left: 0, top: 0 }
			}
		]
	}
];

describe('ZoomCanvasAction', () => {
	test.each<TestCases>(testCases)('Constructor: $name', (tc) => {
		const action: ZoomCanvasAction = new ZoomCanvasAction(tc.options);
		expect(action.zoomCanvasActionOpions).toMatchSnapshot();
	});

	test.each<TestCases>(testCases)('Fire action: $name', (tc) => {
		const action: ZoomCanvasAction = new ZoomCanvasAction(tc.options);
		expect(action.zoomCanvasActionOpions).toMatchSnapshot();

		const engine = new CanvasEngine();
		const cModel = new CanvasModel();
		engine.setModel(cModel);
		action.setEngine(engine);

		expect(cModel.getOffsetX()).toMatchSnapshot();
		expect(cModel.getOffsetY()).toMatchSnapshot();

		const event = generateZoomEvent(tc.events[0]);
		action.options.fire(event);

		expect(cModel.getOffsetX()).toMatchSnapshot();
		expect(cModel.getOffsetY()).toMatchSnapshot();
	});
});
