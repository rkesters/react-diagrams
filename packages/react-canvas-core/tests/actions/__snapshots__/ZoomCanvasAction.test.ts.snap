// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoomCanvasAction Constructor: Options undefined 1`] = `Object {}`;

exports[`ZoomCanvasAction Constructor: Options undefined 2`] = `Object {}`;

exports[`ZoomCanvasAction Constructor: inverseZoom True 1`] = `
Object {
  "inverseZoom": true,
}
`;

exports[`ZoomCanvasAction Constructor: inverseZoom True, ctrlKey true 1`] = `
Object {
  "inverseZoom": true,
}
`;

exports[`ZoomCanvasAction Constructor: inverseZoom false 1`] = `
Object {
  "inverseZoom": false,
}
`;

exports[`ZoomCanvasAction Constructor: inverseZoom false 2`] = `
Object {
  "inverseZoom": false,
}
`;

exports[`ZoomCanvasAction Constructor: inverseZoom undefined 1`] = `Object {}`;

exports[`ZoomCanvasAction Constructor: inverseZoom undefined 2`] = `Object {}`;

exports[`ZoomCanvasAction Fire action: Options undefined 1`] = `Object {}`;

exports[`ZoomCanvasAction Fire action: Options undefined 2`] = `0`;

exports[`ZoomCanvasAction Fire action: Options undefined 3`] = `0`;

exports[`ZoomCanvasAction Fire action: Options undefined 4`] = `-0.03333333333333144`;

exports[`ZoomCanvasAction Fire action: Options undefined 5`] = `-0.03333333333333144`;

exports[`ZoomCanvasAction Fire action: Options undefined 6`] = `Object {}`;

exports[`ZoomCanvasAction Fire action: Options undefined 7`] = `0`;

exports[`ZoomCanvasAction Fire action: Options undefined 8`] = `0`;

exports[`ZoomCanvasAction Fire action: Options undefined 9`] = `-0.03353833333333341`;

exports[`ZoomCanvasAction Fire action: Options undefined 10`] = `-0.03353833333333341`;

exports[`ZoomCanvasAction Fire action: inverseZoom True 1`] = `
Object {
  "inverseZoom": true,
}
`;

exports[`ZoomCanvasAction Fire action: inverseZoom True 2`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom True 3`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom True 4`] = `0.03333333333333258`;

exports[`ZoomCanvasAction Fire action: inverseZoom True 5`] = `0.03333333333333258`;

exports[`ZoomCanvasAction Fire action: inverseZoom True, ctrlKey true 1`] = `
Object {
  "inverseZoom": true,
}
`;

exports[`ZoomCanvasAction Fire action: inverseZoom True, ctrlKey true 2`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom True, ctrlKey true 3`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom True, ctrlKey true 4`] = `0.03353833333333341`;

exports[`ZoomCanvasAction Fire action: inverseZoom True, ctrlKey true 5`] = `0.03353833333333341`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 1`] = `
Object {
  "inverseZoom": false,
}
`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 2`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 3`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 4`] = `-0.03333333333333144`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 5`] = `-0.03333333333333144`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 6`] = `
Object {
  "inverseZoom": false,
}
`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 7`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 8`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 9`] = `-0.03353833333333341`;

exports[`ZoomCanvasAction Fire action: inverseZoom false 10`] = `-0.03353833333333341`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 1`] = `Object {}`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 2`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 3`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 4`] = `-0.03333333333333144`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 5`] = `-0.03333333333333144`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 6`] = `Object {}`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 7`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 8`] = `0`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 9`] = `-0.03353833333333341`;

exports[`ZoomCanvasAction Fire action: inverseZoom undefined 10`] = `-0.03353833333333341`;
